### ISSUE TITLE
Add eclipse-collections support

### ISSUE THREAD

**Issue by yawkat** [FasterXML/jackson-datatypes-collections#29]
Is there interest in support for [eclipse-collections](https://www.eclipse.org/collections/) (primitive collections)? 

I'd be happy to try my hand at implementing support for it if there's no reason against it.

**Comment by cowtowncoder** [comment]
@yawkat The more the merrier, new datatype libraries are always welcome!

I am not familiar with `eclipse-collections`, but if is similar to existing datatypes (as name suggests), it might be easy enough to support.

Whether you want to try to integrate it here immediately, or create a separate repo and integrate later is up to you. Both work, and I can help with details.
If you do want to start straight away here, it would be good to start with `2.9` branch, and we could possibly release 2.9.x versions. 3.0 (which `master` is for) is still quite far out (likely towards end of 2018). It is easy enough to merge from 2.9 to master, although there are differences in codebase that require some work.

**Comment by yawkat** [comment]
I am working on a 2.9 fork of this repo now.

I do have a question about collection types. I have here a class layout that contains both collection and collection-like supertypes:

- InternalIterable extends Iterable and is collection-like (should it be? I'm not sure)
- Bag extends InternalIterable and is definitely collection-like
- MutableCollection extends java.util.Collection and is thus a collection type
- MutableBagIterable extends Bag and MutableCollection (in that order)

The problem with this is that MutableBagIterable extends both the collection-like Bag and the actual collection type MutableCollection. Because of how type construction works (TypeFactory._fromWellKnownInterface) the collection-like supertype Bag is preferred and MutableBagIterable is a collection-like type. This would be fine but isTrueCollectionType on it returns  true when it isn't actually an instance of CollectionType which leads to a classcastexception in DeserializerCache._createCollectionDeserializer2 . Is this a bug or is my class layout / type modifier just broken?

**Comment by yawkat** [comment]
A prototype is available [at my fork](https://github.com/yawkat/jackson-datatypes-collections/tree/eclipse-collections-0)

**Comment by cowtowncoder** [comment]
So, on `TypeModifier`: anything that implements `Map` or `Collection` is typically best left as-is: those will already be "upgraded", as you noticed. There are currently 2 main reasons to add `TypeModifier`, originally to support Scala:

1. `Optional` and similar reference types: ability to recognize these simplifies addition of (de)serializers a lot.
2. Collection-like and Map-like types gets some support in form of automatic resolving of value (and key for map) serializers/deserializers, and polymorphic type wrapping. Not to the degree real Collection/Map types get (standard serializers, possibly deserializer), but half-way there.

In this case I am not sure there is need for type modifier, as it does not sound any of above 2 reasons is relevant?

**Comment by cowtowncoder** [comment]
One other minor observation: for binary types, token deserializer sees may be either `VALUE_STRING` (for textual formats that have no metadata to denote base64-encode; json, xml), or `VALUE_EMBEDDED_OBJECT` for formats that do have metadata to indicate "true" binary (or, in case of YAML, base64-encoded but annotated as binary), such as `protobuf` and `avro`. In both cases `getBinaryValue()` works, but you probably need to check for both.

Another thing that might be of use is that there are a few intermediate base classes above `StdSerializer` / `StdDeserializer` that may come in handy, such as `AsArraySerializerBase` and `ContainerDeserializerBase`. I can't say off-hand if they would work or not, but if they do they can provide some of basic handling, and possibly make easier to get new features added too.

Anyway: good job -- code looks pretty good from what I understand; and if I was more familiar with Eclipse collections I would probably understand it even better. But I trust you are familiar with its handling so it is great that you can work on this.

**Comment by yawkat** [comment]
My collection/collection-like problem goes a little deeper than this. Let me try to be more specific. (this is purely a *de*serializer problem, the normal Iterable serializer works fine for the relevant classes)

The facts leading up to this:

- eclipse-collections includes a number of reference collections. All of these implement the `InternalIterable` interface (an extension to `Iterable`) and `RichIterable` (an extension to `InternalIterable` which adds a `size()`, `toArray()` and so on, making it collection-like in my mind)
- Only *some* of these collections implement `java.util.Collection`. juC includes mutation methods, so only the *mutable* eclipse collections implement juC (for example `MutableBag`).

Now to the problematic part for my implementation:

- My type modifier marks all `RichIterable`s as collection-like so I get automatic element deserializer resolving like you say.
- `MutableBag` implements both `RichIterable` (via `UnsortedBag` -> `Bag` -> `RichIterable`) and `java.util.Collection` (via `MutableBagIterable` -> `MutableCollection` -> juC)
- [`TypeFactory._fromWellKnownInterface`](https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/type/TypeFactory.java#L1322-L1336), when resolving the type of `MutableBag`, prefers the collection-like parent `RichIterable` over `java.util.Collection`. This means that the `JavaType` for `MutableBag` is a `CollectionLikeType`, *not* a `CollectionType` - but because `MutableBag` still implements `java.util.Collection` indirectly, `CollectionLikeType.isTrueCollectionType()` returns true.
- When attempting to construct the deserializer for `MutableBag`, [`DeserializerCache._createDeserializer2`](https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/deser/DeserializerCache.java#L406) errors with a class cast exception. This is because it checks `CollectionLikeType.isTrueCollectionType()` but then casts to `CollectionType` if it returns true - which fails for `MutableBag` for the aforementioned reasons.

There are three fixes I can see:

- Don't have the typemodifier mark eclipse-collections types as collection-like, and implement value deserializer resolving manually. This doesn't seem very nice to me since `RichIterable` and even more so `Bag` are clearly collection-like conceptually.
- Make `TypeFactory` properly resolve `MutableBag` to be a collection type, not just a collection-like type. This sounds like the "right" fix but it is messy to implement, you'd basically have to hardcode an exception that makes `TypeFactory` "prefer" `CollectionType`s over `CollectionLikeType`s.
- Replace the check `isTrueCollectionType` in `DeserializerCache` with a simple `instanceof`. This is the easiest fix but it doesn't feel right either, since `MutableBag` clearly isn't just collection-like, it's an actual `Collection`!

---

[Commit](https://github.com/yawkat/jackson-datatypes-collections/commit/789966fcdcb5bcd9c018cf2e48b4012589daa009)

I have changed the deserializer for byte-based collections to accept `VALUE_EMBEDDED_OBJECT` and treat it the same way. Is this what you meant?

---

I've looked at the subtypes of Std(De)Serializer but found them to be unsuitable or harder to implement. I will try to describe my reasoning (assuming I can still remember it), maybe you can give more suggestions:

- `BaseCollectionDeserializer` implements `StdDeserializer` because this is what `PrimitiveArrayDeserializers` does.
- `BaseCollectionDeserializer` does not implement `ContainerDeserializerBase` because it seemed more suitable for reference types. I did not want to have to deal with content deserializers for primitive collection types. I decided to extend `BaseCollectionDeserializer` for my reference-based collections as well for simplicity. I think this is also what the other deserializers (for example pcollections) in this project do.
- `PrimitiveIterableSerializer` extends `ContainerSerializer` because it seemed suitable. Then again here I *did* produce a content serializer, just null, because that's what the std `ArraySerializerBase` does.
- `PrimitiveIterableSerializer` does not extend `AsArraySerializerBase` because the only thing I see it adding is a bunch of dynamic serializer business that does not seem useful for primitives.

Basically, I followed the std primitive array (de)serializer classes. I'm not sure if they're good examples for this though.

**Comment by nikhilnanivadekar** [comment]
@yawkat thank you very much for this enhancement! It is much appreciated. Please feel free to tag me if you have any Eclipse Collections related questions, will be happy to help!

**Comment by yawkat** [comment]
Okay, thank you.

I've removed my type modifier. All tests pass now. Is [this](https://github.com/yawkat/jackson-datatypes-collections/commit/33e09acaf149f4e2a87e5e4097c7f4706ae78e77#diff-1c7732c297dfe5e18198bf20b3db287cR196) fine for default deserializers and typedeserializers? It passes my tests because the deserializers are contextual but I don't know if this is just luck.

Any other things I need to do before making a PR?

**Comment by yawkat** [comment]
Well I just noticed that the original use case I had for this requires *Map support, so I'll be implementing that as well.

**Comment by yawkat** [comment]
I have implemented deserializer support for maps now ([commit](https://github.com/yawkat/jackson-datatypes-collections/commit/33f68c4c01b44fa6cda82b9360b8380024b6715a)). There are a *lot* of these types (one for each pair of primitives). I did my best to share as much code as possible but there is still a *lot* of code. I generated some of it using a python script - I do not want reflection here (would be annoying for obfuscation / repackaging / etc) and I did not want to bother writing a maven plugin - is there another way or is this fine?

I also changed the compile version to 8 for the module (for toImmutable method refs). I don't think this should be a problem because the eclipse-collections version I am developing against uses java 8 features anyway.

**Comment by yawkat** [comment]
The map serializers are also done now: https://github.com/yawkat/jackson-datatypes-collections/commit/c9b37373fac5963ab912d1f5a726f3465b5d83e1

**Comment by cowtowncoder** [comment]
Thank you for updates -- apologies for slow follow up, I am bit swamped now that I am starting at a new job. I'll go through comments above, answer questions best I can next.

**Comment by cowtowncoder** [comment]
Ok, yes: with primitive types, Java handling, there is no good way around fact there is plenty of duplication.
Java 8 is fine I think, it being a new module and able to pick up initial baseline. Plus Jackson 3.0 is Java 8 anyway.

I think I'd be happy to merge a PR against `2.9` branch of this repo?

One other thing I need (unless I've asked, sorry, so many contributions to review) is CLA:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

(or corporate variant if you prefer, from same repo)

Usual way is to print, fill & sign, scan and email to `info` at fasterxml dot com.
This is needed since it's part of repo with other code, but also make it easier to distribute and accept more contributions (under CLA similarly).
I can also then give you commit access to this module if you would like then -- which means you could also more easily help with other datatypes if you'd like.

**Comment by yawkat** [comment]
I have signed the CLA and mailed it to you.

The PR for these changes is in #30 . I'd be happy with commit access, maybe I'll add support for the few other missing types of eclipse-collections in the future.

**Comment by cowtowncoder** [comment]
Excellent!

**Comment by cowtowncoder** [comment]
Was included in `2.9.5`, although I guess official first minor version will be 2.10.0.
